# model.ipynb

**Projekt:** Spotify-Datensatz  
**Autor:** Erjon Hulaj  
**Datum:** 28.04.2025  

## Einleitung
In diesem Notebook trainiere ich ein maschinelles Lernmodell auf meinem Datensatz, 
um die Popularität von Songs vorherzusagen.  
Dazu teile ich die Daten in Trainings- und Testdaten auf, trainiere ein Modell, 
treffe Vorhersagen und überprüfe deren Qualität.

In [2]:
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

df = pd.read_csv("/Users/erjon/Library/CloudStorage/OneDrive-AlteKantonsschuleAarau/BBB/Module/2. Jahr/M259/LB/dataset.csv")
df.head(10)

Unnamed: 0                track_id                               artists  \
0           0  5SuOikwiRyPMVoIQDJUgSV                           Gen Hoshino   
1           1  4qPNDBW1i3p13qLCt0Ki3A                          Ben Woodward   
2           2  1iJBSr7s7jYXzM8EGcbK5b                Ingrid Michaelson;ZAYN   
3           3  6lfxq3CG4xtTiEg7opyCyx                          Kina Grannis   
4           4  5vjLSffimiIP26QG5WcN2K                      Chord Overstreet   
5           5  01MVOl9KtVTNfFiBU9I7dc                          Tyrone Wells   
6           6  6Vc5wAMmXdKIAM7WUoEb7N  A Great Big World;Christina Aguilera   
7           7  1EzrEOXmMH3G43AXT1y7pA                            Jason Mraz   
8           8  0IktbUcnAGrvD03AWnz3Q8             Jason Mraz;Colbie Caillat   
9           9  7k9GuJYLp2AzqokyEdwEw2                        Ross Copperman   

                                          album_name  \
0                                             Comedy   
1                                   Ghost (Acoustic)   
2                                     To Begin Again   
3  Crazy Rich Asians (Original Motion Picture Sou...   
4                                            Hold On   
5                               Days I Will Remember   
6                        Is There Anybody Out There?   
7                We Sing. We Dance. We Steal Things.   
8                We Sing. We Dance. We Steal Things.   
9                                             Hunger   

                   track_name  popularity  duration_ms  explicit  \
0                      Comedy          73       230666     False   
1            Ghost - Acoustic          55       149610     False   
2              To Begin Again          57       210826     False   
3  Can't Help Falling In Love          71       201933     False   
4                     Hold On          82       198853     False   
5        Days I Will Remember          58       214240     False   
6               Say Something          74       229400     False   
7                   I'm Yours          80       242946     False   
8                       Lucky          74       189613     False   
9                      Hunger          56       205594     False   

   danceability  energy  ...  loudness  mode  speechiness  acousticness  \
0         0.676  0.4610  ...    -6.746     0       0.1430        0.0322   
1         0.420  0.1660  ...   -17.235     1       0.0763        0.9240   
2         0.438  0.3590  ...    -9.734     1       0.0557        0.2100   
3         0.266  0.0596  ...   -18.515     1       0.0363        0.9050   
4         0.618  0.4430  ...    -9.681     1       0.0526        0.4690   
5         0.688  0.4810  ...    -8.807     1       0.1050        0.2890   
6         0.407  0.1470  ...    -8.822     1       0.0355        0.8570   
7         0.703  0.4440  ...    -9.331     1       0.0417        0.5590   
8         0.625  0.4140  ...    -8.700     1       0.0369        0.2940   
9         0.442  0.6320  ...    -6.770     1       0.0295        0.4260   

   instrumentalness  liveness  valence    tempo  time_signature  track_genre  
0          0.000001    0.3580   0.7150   87.917               4     acoustic  
1          0.000006    0.1010   0.2670   77.489               4     acoustic  
2          0.000000    0.1170   0.1200   76.332               4     acoustic  
3          0.000071    0.1320   0.1430  181.740               3     acoustic  
4          0.000000    0.0829   0.1670  119.949               4     acoustic  
5          0.000000    0.1890   0.6660   98.017               4     acoustic  
6          0.000003    0.0913   0.0765  141.284               3     acoustic  
7          0.000000    0.0973   0.7120  150.960               4     acoustic  
8          0.000000    0.1510   0.6690  130.088               4     acoustic  
9          0.004190    0.0735   0.1960   78.899               4     acoustic  

[10 rows x 21 columns]

## (3.1) Aufteilen der Daten in Training und Test

Ich teile meinen Datensatz in Trainingsdaten (80%) und Testdaten (20%) auf.  
Die Features sind wichtige Eigenschaften der Songs (energy, danceability, loudness, valence, duration_ms) 
und die Zielvariable ist popularity.

In [3]:
# Features (X) und Zielvariable (y) definieren
X = df[["energy", "danceability", "loudness", "valence", "duration_ms"]]
y = df["popularity"]

# Aufteilen in Trainings- und Testdaten
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [4]:
# Kontrolle: Aufteilung Trainings-/Testdaten
print(X_train.shape)
print(X_test.shape)

(91200, 5)
(22800, 5)


## (3.2) Auswahl des Algorithmus

Ich habe mich für den Algorithmus **LinearRegression** entschieden, 
weil ich eine kontinuierliche Zielgrösse (**popularity**) vorhersagen möchte. 
Linear Regression ist einfach, schnell und gut interpretierbar. 
Sie eignet sich besonders für Probleme, bei denen numerische Merkmale wie Energy, Danceability oder Loudness 
einen Einfluss auf die Zielvariable haben. 
Für eine erste Modellierung ohne komplexe Zusammenhänge ist sie eine zuverlässige Wahl.

In [5]:
# Modell erstellen und trainieren
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

## (3.3) Vorhersagen treffen und überprüfen

Ich lasse mein Modell Vorhersagen auf den Testdaten machen und vergleiche die Ergebnisse.

In [6]:
# Vorhersagen machen
y_pred = model.predict(X_test)

# Vergleich von tatsächlichen und vorhergesagten Werten
vergleich = pd.DataFrame({
    "Tatsächliche Popularity": y_test.values,
    "Vorhergesagte Popularity": y_pred
})
vergleich.head(10)

Tatsächliche Popularity  Vorhergesagte Popularity
0                       50                 35.339927
1                       11                 34.162657
2                        0                 31.394478
3                        0                 35.568174
4                       57                 31.356086
5                       38                 34.550687
6                       21                 32.159711
7                       54                 34.701781
8                       29                 33.228119
9                       48                 29.113913

### Manuelle Überprüfung einzelner Vorhersagen

Beim Vergleich einzelner Werte fällt auf, dass das Modell z. B. bei Zeile 0 eine tatsächliche Popularität von **50** mit nur **35.3** vorhergesagt hat. Also deutlich zu tief.  
Auch bei Zeile 3 wurde ein Song mit echter Popularität **0** auf **35.5** geschätzt, was unrealistisch ist.  
Das Modell scheint vor allem extrem niedrige und extrem hohe Werte nicht gut zu treffen, sondern alles eher in der Mitte zu schätzen.  
Daraus lässt sich schliessen, dass das Modell **zur Mitte tendiert** und die Popularität nur grob abbildet.

### Erkenntnisse zur Modellqualität

Beim Vergleich der tatsächlichen und vorhergesagten Popularity-Werte zeigt sich, 
dass das Modell grobe Trends gut erkennen kann. 
Die Vorhersagen liegen jedoch oft deutlich unter den tatsächlichen Werten, 
insbesondere bei Songs mit höherer Popularität. 
Dies deutet darauf hin, dass eine einfache lineare Regression 
die Komplexität der Zusammenhänge nur teilweise abbildet. 
Insgesamt erscheinen die Vorhersagen plausibel für eine erste Modellversion, 
lassen aber noch Potenzial für Verbesserungen offen.